vmm/src/device_manager.rs

```diff
@@ -275,6 +276,68 @@ impl Console {
     }
 }
 
+#[cfg(feature = "pci_support")]
+struct VfioDmaMapping {
+    container: Arc<VfioContainer>,
+    memory: Arc<RwLock<GuestMemoryMmap>>,
+}
+
+#[cfg(feature = "pci_support")]
+impl VfioDmaMapping {
+    fn new(container: Arc<VfioContainer>, memory: Arc<RwLock<GuestMemoryMmap>>) -> Self {
+        VfioDmaMapping { container, memory }
+    }
+}
+
+#[cfg(feature = "pci_support")]
+impl ExternalDmaMapping for VfioDmaMapping {
+    fn map(&self, iova: u64, gpa: u64, size: u64) -> result::Result<(), io::Error> {
+        let user_addr = if let Some(addr) = self
+            .memory
+            .read()
+            .unwrap()
+            .get_host_address(GuestAddress(gpa))
+        {
+            addr as u64
+        } else {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                format!(
+                    "failed to convert guest address 0x{:x} into \
+                     host user virtual address",
+                    gpa
+                ),
+            ));
+        };
+
+        self.container
+            .vfio_dma_map(iova, size, user_addr)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::Other,
+                    format!(
+                        "failed to map memory for VFIO container, \
+                         iova 0x{:x}, gpa 0x{:x}, size 0x{:x}: {:?}",
+                        iova, gpa, size, e
+                    ),
+                )
+            })
+    }
+
+    fn unmap(&self, iova: u64, size: u64) -> result::Result<(), io::Error> {
+        self.container.vfio_dma_unmap(iova, size).map_err(|e| {
+            io::Error::new(
+                io::ErrorKind::Other,
+                format!(
+                    "failed to unmap memory for VFIO container, \
+                     iova 0x{:x}, size 0x{:x}: {:?}",
+                    iova, size, e
+                ),
+            )
+        })
+    }
+}
+
```

Comment:
This is all quite VFIO specific, so at the very least it should be part of vfio_device.rs.
And I think it would be cleaner if `ExternalDmaMapping` would be implemented by `VfioDevice` directly.

Reply:
So here you raise a point I ran into. First thing, if we want the VFIO crate to implement this trait, we need VFIO to depend on vm-virtio, but I thought that was not very clean. One solution might be to define an independent crate for this kind of thing.
Second thing, if we directly implement the trait from `VfioDevice`, we need to change the whole ownership of `VfioDevice` into some `Arc<Mutex<VfioDevice>>` since it's currently fully moved to `VfioPciDevice`.

Reply:
> So here you raise a point I ran into. First thing, if we want the VFIO crate to implement this trait, we need VFIO to depend on vm-virtio, but I thought that was not very clean. One solution might be to define an independent crate for this kind of thing.

Yes, that is based on the assumption that those trait should not be defined from vm-virtio. The fact that if we want to have a architecturally cleaner implementation of this trait (i.e. from the VFIO crate itself), then we need to pull this kind of dependencies shows we do need an independent crate. We could use a `vm-device` crate as a placeholder for now.

> Second thing, if we directly implement the trait from `VfioDevice`, we need to change the whole ownership of `VfioDevice` into some `Arc<Mutex<VfioDevice>>` since it's currently fully moved to `VfioPciDevice`. And I want to avoid using Arc<Mutex<>> for this entire device when the only shared part would be the implementation of `ExternalDmaMapping`.